### ISSUE TITLE
Code fields have autocapitalizations enabled on mobile

### ISSUE THREAD

**Issue by roger-** [jupyter/notebook#539]
In Safari mobile on https://try.jupyter.org, entering text in a new code field automatically capitalizes the first character. 

I believe this and autocorrect should be disabled, as shown [here](http://stackoverflow.com/questions/5171764/how-do-you-turn-off-auto-capitalisation-in-html-form-fields-in-ios).

**Comment by jdfreder** [comment]
I don't think those attributes, that are shown in the stack overflow post you've shared, map one to one with CodeMirror (the component we use to render the code field).  CodeMirror may not support this, but it's worth looking into.

I should also note, we don't "officially" support Safari mobile nor any other mobile platform.  That's not to say we shouldn't try though!

**Comment by roger-** [comment]
Thanks but it seems that CodeMirror should already do this: https://groups.google.com/forum/m/#!topic/codemirror/n6CRZ3CUO3s

Not sure why it apparently isn't the case.

**Comment by jasongrout** [comment]
I was going to post that we had the same discussion with CodeMirror years ago for Sage, but @roger beat me to it.  That was a previous version of codemirror, though, and its design was completely different and it's seen at least two major rewrites since then.

**Comment by 8leggedunicorn** [comment]
@jdfreder @jasongrout @roger- I looked at the current state of the code and it appears there are explicit declarations in Jupyter notebook turning off auto-capitalization:

```
➜  notebook git:(master) find -name '*.js' | while read file ; do grep autocapitalize "$file" ; done
➜  notebook git:(master) find -name '*.js' | while read file ; do grep -H autocapitalize "$file" ; done
./notebook/static/tree/js/main.min.js:    field.setAttribute("autocapitalize", "off");
./notebook/static/edit/js/main.min.js:    field.setAttribute("autocapitalize", "off");
./notebook/static/notebook/js/main.min.js:    field.setAttribute("autocapitalize", "off");
grep: ./notebook/static/components/term.js: Is a directory
./notebook/static/components/term.js/src/term.js:  textarea.autocapitalize = 'none';
./notebook/static/components/codemirror/lib/codemirror.js:    field.setAttribute("autocapitalize", "off");
./notebook/static/terminal/js/main.min.js:  textarea.autocapitalize = 'none';
./notebook/static/terminal/js/main.min.js:    field.setAttribute("autocapitalize", "off");
```
A notable caveat is that the code being used right now has the value 'off' for `autocapitalize`, however looking at the [developer manual for Safari Apple](https://developer.apple.com/library/safari/documentation/AppleApplications/Reference/SafariHTMLRef/Articles/Attributes.html) this is deprecated in favor of using the value 'none' since iOS 5.0.  The [equivalent state](https://developers.google.com/web/updates/2015/04/autocapitalize?hl=en) appears to be in recent versions of Android Chrome. I don't know enough about this but a simple PR changing 'off' to 'none' maybe all that's necessary to rectify this issue, but well as you know many of those JavaScript files are thousands of loc so who knows, but it might be worth a try.

@roger- This being the case note the mobile support is still at this time experimental in CodeMirror, and was only officially experimentally supported February of this year, note the [release history](http://codemirror.net/doc/releases.html) for version 5.0.  [This](http://codemirror.net/#browsersupport) is the list of their officially supported browsers, all desktop versions. 

Secondly if one does a [search on the GitHub repository of CodeMirror for 'keyboard'](https://github.com/codemirror/CodeMirror/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+keyboard) there are currently 7 open issues 5 of which are mobile platform related. 

The project appears to be actively working on becoming more compatible with mobile browsers, and I will certainly benefit as my portable device of choice is an Android tablet. Note at least on my Android tablet the hardware (bluetooth) keyboard I use with it doesn't auto-capitalize the first inputted letter like the virtual keyboard does.  For my and potentially your workflow using a hardware keyboard is a reasonable stopgap, but of course what mobile OS, what OS version, as well as browser used on each may radically change the actual behavior observed between different mobile setups.

**Comment by jdfreder** [comment]
> Edit: Ah these JavaScript files are being generated in the build process. Someone more knowledgeable than I will likely need to find where.


The main.min.js files contain a bundle of all of the Javascript dependencies for the given page.  We use r.js to traverse the dependency tree and do the bundling.  The code that you point to probably lives in CodeMirror's source, because searching the notebook repository for it yields no results.

**Comment by jasongrout** [comment]
Yes, the line is here in codemirror: https://github.com/codemirror/CodeMirror/blob/5fafcf837480430b614085d62d884098ccd53297/lib/codemirror.js#L1182

**Comment by paulmasson** [comment]
CodeMirror currently displays code in a `div` with `contentEditable='true'`. The fixes on Stackoverflow from the original poster don't work in this context. CodeMirror acknowledges this:

https://github.com/codemirror/CodeMirror/issues/3087

I had hoped to find a way to fix this for SageMathCell, but no go.

One can always turn off autocapitalization completely on the mobile device.

**Comment by jbbqqf** [comment]
Hi! Triaging older issues this week. Looks like this one can probably be closed here: the mobile-input behavior of code fields is now governed by JupyterLab's editor, not this repo.

- Notebook v7 is rebuilt on JupyterLab components — "a significant change to the `jupyter/notebook` code base" ([README.md L21-L30](https://github.com/jupyter/notebook/blob/main/README.md#L21-L30)).
- Classic-frontend code-input fields (the ones with the autocapitalize bug at try.jupyter.org back in 2015) live in [`jupyter/nbclassic`](https://github.com/jupyter/nbclassic) for v6 maintenance ([README.md L34-L36](https://github.com/jupyter/notebook/blob/main/README.md#L34-L36)).

CodeMirror (the editor under JupyterLab cells in v7) controls those `autocapitalize` / `autocorrect` attributes on its `<textarea>`/`contenteditable` host. If the issue still reproduces on a recent mobile Safari against v7, `jupyterlab/jupyterlab` is the right venue. Would you mind closing here? Happy to be wrong if I missed a v7-side hook.

(Disclosure: I drafted this comment with help from Claude Code while triaging stale issues; the references above were verified manually against the current `main` branch.)

**Comment by jtpio** [comment]
Thanks @jbbqqf!

Indeed, this would be something to address in JupyterLab if still relevant. Closing.
